climada/entity/exposures/test/test_base.py

```diff
@@ -89,6 +90,20 @@ def test__init__meta_type(self):
         self.assertEqual("meta must be a dictionary",
                       str(cm.exception))
 
+    def test__init__geometry_type(self):
+        """Check that initialization fails when `geometry` is given as a `str` argument"""
+        with self.assertRaises(ValueError) as cm:
+            Exposures(geometry='myname')
+        self.assertEqual("Exposures is not able to handle customized 'geometry' column names.",
+                         str(cm.exception))
+
+    def test__init__mda_in_kwargs(self):
+        """Check if mda is not present in kwargs after being removed"""
+        litpop = LitPop(exponents=2)
+        self.assertEqual(litpop.exponents, 2)
+        litpop = LitPop(meta=dict(exponents=3))
+        self.assertEqual(litpop.exponents, 3)
```

Comment:
much better now 😉 though I still cannot make sense out of the test description. What is "mda", where is it removed and how is that related to the the `exponents`?

Reply:
This is due to the `Entity` class being written quite badly, if I may say so. We try to test [`climada/entity/exposures/base.py#L178`](https://github.com/CLIMADA-project/climada_python/blob/main/climada/entity/exposures/base.py#L178) and following. There is a class attribute `_metadata` which is supposed to be extended by derived classes. The base class `__init__`, however, is built to handle these extensions. L178 and L179 check if there are keywords in the *current* class type `_metadata` that is not part of the `Exposures._metadata`. This can only be the case if `self` is a derived class, and not `Exposures` itself. The following code will then extract these keywords from the `kwargs` before the remainder of the `kwargs` is passed to the GeoDataframe.

See [`Exposures._metadata`](https://github.com/CLIMADA-project/climada_python/blob/main/climada/entity/exposures/base.py#L125) and [`LitPop._metadata`](https://github.com/CLIMADA-project/climada_python/blob/main/climada/entity/exposures/litpop/litpop.py#L69)

This is all very convoluted, and the only advantage that I see is that you do not need a dedicated `__init__` for derived classes anymore, at the expense of code that is very difficult to understand.

Reply:
Oh, what's even worse is that the `Exposures.__init__` will thus instantiate attributes with names listed in `_metadata`, even for derived classes.

Reply:
I see. Then the suggested description would about describe the purpose of this test?